worker/download.go

```diff
@@ -0,0 +1,1088 @@
+		span.SetAttributes(attribute.Int64("duration", elapsed.Milliseconds()))
+		span.RecordError(err)
+		span.End()
+	}()
+
+	// download the sector
+	buf := bytes.NewBuffer(make([]byte, 0, rhpv2.SectorSize))
+	err = d.host.DownloadSector(req.ctx, buf, req.root, req.offset, req.length)
+	if err != nil {
+		req.fail(err)
+		return err
+	}
+
+	d.mu.Lock()
+	d.numDownloads++
+	d.mu.Unlock()
+
+	req.succeed(buf.Bytes())
+	return nil
+}
+
+func (req *sectorDownloadReq) succeed(sector []byte) {
+	select {
+	case <-req.ctx.Done():
+	case req.responseChan <- sectorDownloadResp{
+		hk:          req.hk,
+		sectorIndex: req.sectorIndex,
+		sector:      sector,
+	}:
+	}
+}
+
+func (req *sectorDownloadReq) fail(err error) {
+	select {
+	case <-req.ctx.Done():
+	case req.responseChan <- sectorDownloadResp{
+		hk:  req.hk,
+		err: err,
+	}:
+	}
+}
+
+func (req *sectorDownloadReq) done() bool {
+	select {
+	case <-req.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *slabDownload) overdrive(ctx context.Context, respChan chan sectorDownloadResp) (resetTimer func()) {
+	// overdrive is disabled
+	if s.mgr.overdriveTimeout == 0 {
+		return func() {}
+	}
+
+	// create a timer to trigger overdrive
+	timer := time.NewTimer(s.mgr.overdriveTimeout)
+	resetTimer = func() {
+		timer.Stop()
+		select {
+		case <-timer.C:
+		default:
+		}
+		timer.Reset(s.mgr.overdriveTimeout)
+	}
+
+	// create a function to check whether overdrive is possible
+	canOverdrive := func() bool {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		// overdrive is not due yet
+		if time.Since(s.lastOverdrive) < s.mgr.overdriveTimeout {
+			return false
+		}
+
+		// overdrive is maxed out
+		remaining := s.minShards - s.numCompleted
+		if s.numInflight >= s.mgr.maxOverdrive+uint64(remaining) {
+			return false
+		}
+
+		s.lastOverdrive = time.Now()
+		return true
+	}
+
+	// try overdriving every time the timer fires
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				if canOverdrive() {
+					req := s.nextRequest(ctx, respChan, true)
+					if req == nil {
+						fmt.Printf("DEBUG PJ: %v | %d | no sector to overdrive\n", s.dID, s.index)
+					} else {
+						lErr := s.launch(req) // ignore error
+						fmt.Printf("DEBUG PJ: %v | %d | overdriving sector %d err %v\n", s.dID, s.index, req.sectorIndex, lErr)
+					}
+				}
+				resetTimer()
+			}
+		}
+	}()
+
+	return
+}
+
+func (s *slabDownload) nextRequest(ctx context.Context, responseChan chan sectorDownloadResp, overdrive bool) *sectorDownloadReq {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// prepare next sectors to download
+	if len(s.hostToSectors[s.curr]) == 0 {
+		// grab unused hosts
+		var hosts []types.PublicKey
+		for host := range s.hostToSectors {
+			if _, used := s.used[host]; !used {
+				hosts = append(hosts, host)
+			}
+		}
+
+		// sort them and assign the current host
+		s.mgr.sort(hosts)
+		for _, host := range hosts {
+			s.used[host] = struct{}{}
+			s.curr = host
+			break
```

Comment:
Sorting is quite excessive considering that we are only interested in finding the fastest host which is a faster operation than a full sort. Maybe rename `sort` to `fastest` and return the fastest host from the list?

Reply:
Haha, never considered this. Good stuff 👍